CDAP-6584 add kafka streaming source

albertshau opened the pull request.

albertshau committed:
CDAP-6584 add kafka streaming source

gokulavasan commented on spark-plugins/src/main/java/co/cask/hydrator/plugin/batch/spark/KafkaStreamingSource.java:
```diff
@@ -0,0 +1,184 @@
+  /**
+   * Conf for Kafka streaming source.
+   */
+  public static class Conf extends PluginConfig {
```
Need to use `ReferencePluginConfig` instead of `PluginConfig` ?

albertshau replied:
in order to use it, StreamingSource needs some additional capabilities. Opened HYDRA-372.

sreevatsanraman commented on spark-plugins/docs/Kafka-streamingsource.md:
```diff
@@ -0,0 +1,70 @@
+
+    {
+        "name": "Kafka",
+        "type": "realtimesource",
```
streamingsource?

albertshau committed:
HYDRA-177 source enhancements

Adds reference name in order to provide tracking for the kafka
brokers and topics configured in the source. Also adding config
properties that allow users to include the message key and the
batch timestamp in the output records.

gokulavasan commented on spark-plugins/src/main/java/co/cask/hydrator/plugin/batch/spark/ReferenceStreamingSource.java:
```diff
@@ -0,0 +1,65 @@
+    pipelineConfigurer.createDataset(conf.referenceName, Constants.EXTERNAL_DATASET_TYPE, DatasetProperties.EMPTY);
+  }
+
+  protected void registerUsage(Transactional transactional) throws TransactionFailureException {
```
Where is this method invoked?

albertshau replied:
ah forgot to actually call it...

gokulavasan commented on spark-plugins/docs/Kafka-streamingsource.md:
```diff
@@ -0,0 +1,92 @@
+            "format": "csv",
+            "timeField": "readTime",
+            "keyField": "key",
+            "schema": "{
```
Will we be adding offset to read from later? Also, will we add the ability to persist offset?

albertshau replied:
yeah, though that is dependent on HYDRA-378, in which case the kafka dstream library will store offsets in the spark checkpoints on its own.

gokulavasan commented:
Couple of questions. Rest LGTM 👍
